jetty-util/src/main/java/org/eclipse/jetty/util/thread/strategy/AdaptiveExecutionStrategy.java

```diff
@@ -0,0 +1,482 @@
+                    break;
+            }
+        }
+        if (LOG.isDebugEnabled())
+            LOG.debug("{} dispatch {}", this, execute);
+        if (execute)
+            _executor.execute(this);
+    }
+
+    @Override
+    public void run()
+    {
+        tryProduce(true);
+    }
+
+    @Override
+    public void produce()
+    {
+        tryProduce(false);
+    }
+
+    private void tryProduce(boolean wasPending)
+    {
+        if (LOG.isDebugEnabled())
+            LOG.debug("{} tryProduce {}", this, wasPending);
+
+        try (AutoLock l = _lock.lock())
+        {
+            if (wasPending)
+                _pending = false;
+
+            switch (_state)
+            {
+                case IDLE:
+                    // Enter PRODUCING
+                    _state = State.PRODUCING;
+                    break;
+
+                case PRODUCING:
+                    // Keep other Thread producing
+                    _state = State.REPRODUCING;
+                    return;
+
+                default:
+                    return;
+            }
+        }
+
+        boolean nonBlocking = Invocable.isNonBlockingInvocation();
+
+        while (isRunning())
+        {
+            try
+            {
+                if (doProduce(nonBlocking))
+                    continue;
+                return;
+            }
+            catch (Throwable th)
+            {
+                LOG.warn("Unable to produce", th);
+            }
+        }
+    }
+
+    private boolean doProduce(boolean nonBlocking)
+    {
+        Runnable task = produceTask();
+
+        if (task == null)
+        {
+            try (AutoLock l = _lock.lock())
+            {
+                // Could another task just have been queued with a produce call?
+                switch (_state)
+                {
+                    case PRODUCING:
+                        _state = State.IDLE;
+                        return false;
+
+                    case REPRODUCING:
+                        _state = State.PRODUCING;
+                        return true;
+
+                    default:
+                        throw new IllegalStateException(toStringLocked());
+                }
+            }
+        }
+
+        Mode mode;
+        Invocable.InvocationType taskType = Invocable.getInvocationType(task);
+        if (taskType == Invocable.InvocationType.NON_BLOCKING)
+        {
+            // The produced task will never block, so we directly consume the task with PC
+            // and then resume production.
+            mode = Mode.PRODUCE_CONSUME;
+        }
+        else
+        {
+            // The produced task might block. Is this producing thread allowed to block?
+            boolean mayBlock = !nonBlocking;
+
+            // only if this thread may block do we need to take the lock so we can
+            // atomically check/mutate the state and pending status.
+            try (AutoLock l = mayBlock ? _lock.lock() : AutoLock.NO_LOCK)
```

Comment:
Please no.
Can we just revert the algorithm as it was before? 
Much simpler, much more readable, each case had its own block and it was easy to reason.

Reply:
I don't agree. In the previous algorithm, each Sun strategy is selected multiple times in multiple locations over several pages of code, in a store that didn't at all match the documentation.

I'll revert to the version without the lock back, but dinner duplication of code... but with comments